pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,233 @@
+						Namespace: r.Namespace,
+					},
+				}
+				requests = append(requests, request)
+
+				// Add the Router Name to the currentIngressSet.
+				currentRouteIngressSet.Insert(ri.RouterName)
+			}
+		}
+	}
+
+	// Get the previous set of Ingresses of the Route.
+	previousRouteIngressSet := r.routeIngressMap[routeNamespacedName]
+
+	// Iterate through the previousRouteIngressSet.
+	for routerName := range previousRouteIngressSet {
+		// Check if the currentRouteIngressSet contains the Router Name. If it does not,
+		// then the Ingress was removed from the Route Status. The reconcile loop is needed
+		// to be run for the corresponding Ingress Controller.
+		if !currentRouteIngressSet.Has(routerName) {
+			log.Info("queueing ingresscontroller", "name", routerName)
+			// Create a reconcile.Request for the router named in the RouteIngress.
+			request := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      routerName,
+					Namespace: r.Namespace,
+				},
+			}
+			requests = append(requests, request)
+		}
+	}
+
+	// Map the currentRouteIngressSet to Route's NamespacedName.
+	r.routeIngressMap[routeNamespacedName] = currentRouteIngressSet
+
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache           cache.Cache
+	Namespace       string
+	routeIngressMap map[types.NamespacedName]sets.String
+}
+
+// Reconcile expects request to refer to an Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// If the ingresscontroller is deleted, then delete the corresponding RouteMetricsControllerRoutesPerShard metric label
+	// and return early.
+	if ingressController.DeletionTimestamp != nil {
+		// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+		DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+		log.Info("RoutesPerShard metric label corresponding to the Ingress Controller is successfully deleted", "ingresscontroller", ingressController)
+		return reconcile.Result{}, nil
+	}
+
+	// NOTE: Even though the route admitted status should reflect validity of the namespace labelselector, we still will validate
```

Comment:
Thanks for the update regarding the bug.

I have actually updated the comment to include route labelselector as well. This is the current comment:
```
// NOTE: Even though the route admitted status should reflect validity of the namespace and route labelselectors, we still will validate
// the namespace and route labels as there are still edge scenarios where the route status may be inaccurate.
```

Reply:
Whoops, sorry I think my comment may have been misunderstood.

We think you should only use `routeAdmitted`, and not check namespaces or route labels, with the understanding we need to fix https://issues.redhat.com/browse/OCPBUGS-1689 in order to solve the edge cases.

Does that make sense? I think you can delete a lot of code here and make it much more simpler.

Reply:
Oh got it.

But fetching the routes based on the route and namespace labelselectors would actually reduce the number of iteration over the list of routes. Or else in every reconcile we would have to check the status of every route. I actually wasn't aware of the edge case issues of the route status not being cleared when I had written the code. Would we not want to keep the optimized code despite the bug being there or not?

If you strongly feel that the code is complicated then I can remove the fetching of the routtes based on the labelselectors.

Reply:
That's a good point. I see you would need to iterate over all routes if you removed the logic to filter on namespace and route selector. This is a tricky one. I agree, the way the code is written, it's more efficient to just filter via namespace and router label selectors. I think we should still use `routeAdmitted`, but consider changing this to `routeStatusAdmitted` to be clear.

Given that, I think we attempt to de-duplicate this logic and collapse into a common function since we mostly duplicating the logic in https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/router_status.go#L143-L169. If we have a change in shard labeling logic in the future, we probably only want to change it in one place. Also, I think you did it more efficiently than I did it by listing the route with the route selector and I listed all routes then filtered by the route selector. So I'd prefer to use your logic.

I'd suggest the interface for the common function in `router_status.go` to be something like:
```
func RoutesAdmittedByIngress(c client.Reader, ingress *operatorv1.IngressController) *routev1.RouteList {
```
Couple of notes if you plan to do this:
1. I use the client, not cache, due to a race condition, hence the c.clientReader interface
2. You'll need to so some funky importing, I'm not sure of the best way, but I think something like `import ingress "github.com/openshift/cluster-ingress-operator/pkg/operator/controller/ingress"` would work.

Let me know if this makes sense, and you think we should try to de-duplicate some of this logic between me and you. Thanks for being responsive on this, I know the route status stuff is messy.

Reply:
It would be good to de-duplicate the logic. However, the `RouteList` differs in the 2 scenarios. In the `route-metrics-controller` scenario, the list of routes are required which matches the Namespace and Route labelselectors and also has the `admitted` status, whereas in the scenario of the `clearRoutesNotAdmittedByIngress` function in `router_status.go`, the list of routes are required which doesn't match the Namespace and Route labelselectors but has the `admitted` status.


> I think you did it more efficiently than I did it by listing the route with the route selector and I listed all routes then filtered by the route selector.

In your case you have to list all the routes as you are looking for the routes which doesn't match the Namespace and Route labelselectors but has the `admitted` status. If you fetch using the Route labelselector then you'll not be able to clear the status of those routes which has `admitted` status but does not match the  Route labelselector.

One possible way to de-duplicate the logic is to have a single function which returns both the list of routes which matches and does not match the labelselectors. Now in this scenario we cannot use any labelselector to fetch the routes as we have to also get the routes which doesn't match the labelselectors. So ultimately we are fetching all routes anyway and then filtering them out.

Now it makes sense to me to fetch all the routes without using and labelselectors and just rely on the status of each of them for gathering the metrics. Otherwise, any change in labeling logic would need changes in multiple places. PLMK your thoughts.

The route status stuff is really messy :smiling_face_with_tear:

Reply:
> I think we should still use `routeAdmitted`, but consider changing this to `routeStatusAdmitted` to be clear.

Changed the function name.